src/main/java/picard/illumina/IlluminaBasecallsToFastq.java

```diff
@@ -632,77 +508,138 @@ public void close() {
-        private final int numSampleBarcodes;
-        private final int numMolecularBarcodes;
-
-        private BasicFastqWriter writer = null;
-        private FastqReader reader = null;
-
-        FastqRecordsForClusterCodec(final int numTemplates, final int numSampleBarcodes, final int numMolecularBarcodes) {
-            this.numTemplates = numTemplates;
-            this.numSampleBarcodes = numSampleBarcodes;
-            this.numMolecularBarcodes = numMolecularBarcodes;
-        }
+    private static class ClusterDataCodec implements SortingCollection.Codec<ClusterData> {
+        private final BinaryCodec binaryCodec = new BinaryCodec();
 
         @Override
-        public void setOutputStream(final OutputStream os) {
-            writer = new BasicFastqWriter(new PrintStream(os));
+        public void setOutputStream(OutputStream os) {
+            this.binaryCodec.setOutputStream(os);
         }
 
         @Override
-        public void setInputStream(final InputStream is) {
-            reader = new FastqReader(new BufferedReader(new InputStreamReader(is)));
+        public void setInputStream(InputStream is) {
+            this.binaryCodec.setInputStream(is);
         }
 
-        //TODO: add tests to encode and decode
         @Override
-        public void encode(final FastqRecordsForCluster val) {
-            if (numTemplates != val.templateRecords.length) throw new IllegalStateException();
-            if (numSampleBarcodes != val.sampleBarcodeRecords.length) throw new IllegalStateException();
-            encodeArray(val.templateRecords);
-            encodeArray(val.sampleBarcodeRecords);
-            encodeArray(val.molecularBarcodeRecords);
-            writer.flush();
+        public void encode(ClusterData clusterData) {
+            this.binaryCodec.writeInt(clusterData.getLane());
+            this.binaryCodec.writeInt(clusterData.getTile());
+            this.binaryCodec.writeInt(clusterData.getX());
+            this.binaryCodec.writeInt(clusterData.getY());
+            this.binaryCodec.writeInt(clusterData.getNumReads());
+
+            for (int i = 0; i < clusterData.getNumReads(); i++) {
+                ReadData read = clusterData.getRead(i);
+                byte[] bases = read.getBases();
+                byte[] quals = read.getQualities();
+
+                binaryCodec.writeInt(bases.length);
+                binaryCodec.writeString(read.getReadType().name(), false, false);
+
+                for (int j = 0; j < bases.length; j++) {
+                    this.binaryCodec.writeByte(bases[j]);
+                    this.binaryCodec.writeByte(quals[j]);
+                }
+            }
+
+            this.binaryCodec.writeBoolean(clusterData.isPf());
+            if (clusterData.getMatchedBarcode() != null) {
+                this.binaryCodec.writeString(clusterData.getMatchedBarcode(), true, true);
+            } else {
+                this.binaryCodec.writeString("", true, false);
+            }
```

Comment:
Unless there's a good reason I think it would be nicer to write all the 1-per-cluster items first before the reads, and then the reads.